File: fashion_mnist_cnn.R
# Fashion MNIST Classification using CNN in R
# Load required libraries
library(keras)
library(tensorflow)
library(tidyverse)
library(caret)
library(ggplot2)

# Set random seed for reproducibility
set.seed(42)
tensorflow::tf$random$set_seed(42)

# 1. Load the Fashion MNIST dataset
cat("Loading Fashion MNIST dataset...\n")
fashion_mnist <- dataset_fashion_mnist()

# Split into training and test sets
c(train_images, train_labels) %<-% fashion_mnist$train
c(test_images, test_labels) %<-% fashion_mnist$test

# Get dataset information
cat(sprintf("Training data shape: %s x %s\n", dim(train_images)[1], paste(dim(train_images)[2:3], collapse=" x ")))
cat(sprintf("Test data shape: %s x %s\n", dim(test_images)[1], paste(dim(test_images)[2:3], collapse=" x ")))

# Define class names for better visualization
class_names <- c('T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
                 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot')

# 2. Data Preprocessing
# Normalize pixel values to be between 0 and 1
train_images <- train_images / 255
test_images <- test_images / 255

# Reshape the data to fit the model (add channel dimension)
train_images <- array_reshape(train_images, c(dim(train_images), 1))
test_images <- array_reshape(test_images, c(dim(test_images), 1))

# Convert labels to categorical
train_labels_categorical <- to_categorical(train_labels, 10)
test_labels_categorical <- to_categorical(test_labels, 10)

# 3. Build a 6-layer CNN model
model <- keras_model_sequential() %>%
  # Layer 1: Convolutional Layer
  layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = 'relu',
                input_shape = c(28, 28, 1), padding = 'same') %>%
  
  # Layer 2: Max Pooling Layer
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%
  
  # Layer 3: Convolutional Layer
  layer_conv_2d(filters = 64, kernel_size = c(3, 3), activation = 'relu',
                padding = 'same') %>%
  
  # Layer 4: Max Pooling Layer
  layer_max_pooling_2d(pool_size = c(2, 2)) %>%
  
  # Layer 5: Fully Connected Layer
  layer_flatten() %>%
  layer_dense(units = 128, activation = 'relu') %>%
  layer_dropout(rate = 0.5) %>%
  
  # Layer 6: Output Layer
  layer_dense(units = 10, activation = 'softmax')

# Compile the model
model %>% compile(
  optimizer = 'adam',
  loss = 'categorical_crossentropy',
  metrics = c('accuracy')
)

# Display model summary
summary(model)

# 4. Train the model
cat("\nTraining the model...\n")
history <- model %>% fit(
  x = train_images,
  y = train_labels_categorical,
  epochs = 10,
  batch_size = 64,
  validation_split = 0.2,
  verbose = 1
)

# 5. Evaluate the model
cat("\nEvaluating the model...\n")
scores <- model %>% evaluate(test_images, test_labels_categorical, verbose = 0)
cat("Test loss:", scores[[1]], "\n")
cat("Test accuracy:", scores[[2]], "\n")

# 6. Plot training history
png("training_history_r.png", width = 1200, height = 500, res = 100)
par(mfrow = c(1, 2))
plot(history$metrics$accuracy, type = "l", col = "blue", 
     xlab = "Epoch", ylab = "Accuracy", main = "Model Accuracy")
lines(history$metrics$val_accuracy, col = "red")
legend("bottomright", legend = c("Training", "Validation"), 
       col = c("blue", "red"), lty = 1)

plot(history$metrics$loss, type = "l", col = "blue", 
     xlab = "Epoch", ylab = "Loss", main = "Model Loss")
lines(history$metrics$val_loss, col = "red")
legend("topright", legend = c("Training", "Validation"), 
       col = c("blue", "red"), lty = 1)
dev.off()

# 7. Make predictions for specific test images
cat("\nMaking predictions for specific images...\n")

# Function to visualize predictions for a specific image
predict_specific_image <- function(image_idx) {
  # Get the image and reshape it
  image <- array_reshape(test_images[image_idx,,,,drop=FALSE], c(1, 28, 28, 1))
  true_label <- test_labels[image_idx] + 1  # Add 1 for R indexing
  
  # Make prediction
  prediction <- model %>% predict(image)
  predicted_class <- which.max(prediction[1,])
  confidence <- prediction[1, predicted_class] * 100
  
  # Create visualization
  png(paste0("specific_prediction_r_", image_idx, ".png"), width = 600, height = 600, res = 100)
  par(mar = c(1, 1, 4, 1))
  image_to_plot <- array_reshape(image, c(28, 28))
  image(t(apply(image_to_plot, 2, rev)), axes = FALSE, col = gray.colors(100), 
        main = paste0("Prediction: ", class_names[predicted_class], 
                      " (", round(confidence, 2), "%)\n",
                      "True: ", class_names[true_label]))
  dev.off()
  
  # Return results
  result <- list(
    image_idx = image_idx,
    true_class = class_names[true_label],
    predicted_class = class_names[predicted_class],
    confidence = confidence,
    correct = predicted_class == true_label
  )
  
  return(result)
}

# Predict two specific images
image_indices <- c(42, 123)
results <- list()

for (idx in image_indices) {
  result <- predict_specific_image(idx)
  results[[length(results) + 1]] <- result
  cat(paste0("Image ", idx, ":\n"))
  cat(paste0("  True class: ", result$true_class, "\n"))
  cat(paste0("  Predicted class: ", result$predicted_class, "\n"))
  cat(paste0("  Confidence: ", round(result$confidence, 2), "%\n"))
  cat(paste0("  Correct: ", result$correct, "\n\n"))
}

# 8. Create confusion matrix
cat("\nGenerating confusion matrix...\n")
y_pred <- model %>% predict(test_images)
y_pred_classes <- max.col(y_pred) - 1  # Convert to 0-based indexing
conf_matrix <- confusionMatrix(factor(y_pred_classes), factor(test_labels))

# Save confusion matrix visualization
library(lattice)
png("confusion_matrix_r.png", width = 1200, height = 1000, res = 100)
levelplot(conf_matrix$table, 
          main = "Confusion Matrix",
          xlab = "Predicted", ylab = "Actual",
          col.regions = colorRampPalette(c("white", "blue"))(100))
dev.off()

print(conf_matrix)

# Save the model
model %>% save_model_hdf5("fashion_mnist_cnn_model_r.h5")
cat("Model saved as 'fashion_mnist_cnn_model_r.h5'\n")

cat("\nR implementation completed successfully!\n")

File: fashion_mnist_cnn.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf
keras = tf.keras
layers = tf.keras.layers
models = tf.keras.models
fashion_mnist = tf.keras.datasets.fashion_mnist
to_categorical = tf.keras.utils.to_categorical
from sklearn.metrics import classification_report, confusion_matrix
import seaborn as sns

# Set random seed for reproducibility
np.random.seed(42)
tf.random.set_seed(42)

# 1. Load the Fashion MNIST dataset
print("Loading the Fashion MNIST dataset...")
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# Get dataset information
print(f"Training data shape: {X_train.shape}")
print(f"Test data shape: {X_test.shape}")

# 2. Data Preprocessing
# Normalize pixel values to be between 0 and 1
X_train = X_train.astype('float32') / 255.0
X_test = X_test.astype('float32') / 255.0

# Reshape the data to fit the model (28x28x1)
X_train = X_train.reshape(-1, 28, 28, 1)
X_test = X_test.reshape(-1, 28, 28, 1)

# One-hot encode the target variable
y_train_categorical = to_categorical(y_train, num_classes=10)
y_test_categorical = to_categorical(y_test, num_classes=10)

# Define class names for better visualization
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

# 3. Visualize some examples from the dataset
plt.figure(figsize=(10, 10))
for i in range(25):
    plt.subplot(5, 5, i + 1)
    plt.xticks([])
    plt.yticks([])
    plt.grid(False)
    plt.imshow(X_train[i].reshape(28, 28), cmap=plt.cm.binary)
    plt.xlabel(class_names[y_train[i]])
plt.suptitle('Fashion MNIST Examples', fontsize=16)
plt.savefig('fashion_mnist_examples.png', dpi=300, bbox_inches='tight')
plt.close()


# 4. Build a 6-layer CNN model
def create_model():
    model = models.Sequential([
        # Layer 1: Convolutional Layer
        layers.Conv2D(32, (3, 3), activation='relu', padding='same', input_shape=(28, 28, 1)),

        # Layer 2: Max Pooling Layer
        layers.MaxPooling2D((2, 2)),

        # Layer 3: Convolutional Layer
        layers.Conv2D(64, (3, 3), activation='relu', padding='same'),

        # Layer 4: Max Pooling Layer
        layers.MaxPooling2D((2, 2)),

        # Layer 5: Fully Connected Layer
        layers.Flatten(),
        layers.Dense(128, activation='relu'),
        layers.Dropout(0.5),  # Add dropout to prevent overfitting

        # Layer 6: Output Layer
        layers.Dense(10, activation='softmax')
    ])

    # Compile the model
    model.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    return model


# Create and summarize the model
model = create_model()
model.summary()

# 5. Train the model
print("\nTraining the model...")
history = model.fit(
    X_train, y_train_categorical,
    epochs=10,
    batch_size=64,
    validation_split=0.2,
    verbose=1
)

# 6. Evaluate the model
print("\nEvaluating the model...")
test_loss, test_acc = model.evaluate(X_test, y_test_categorical, verbose=2)
print(f"Test accuracy: {test_acc:.4f}")

# 7. Visualize training history
plt.figure(figsize=(12, 5))

# Plot training & validation accuracy
plt.subplot(1, 2, 1)
plt.plot(history.history['accuracy'])
plt.plot(history.history['val_accuracy'])
plt.title('Model Accuracy')
plt.ylabel('Accuracy')
plt.xlabel('Epoch')
plt.legend(['Train', 'Validation'], loc='lower right')

# Plot training & validation loss
plt.subplot(1, 2, 2)
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('Model Loss')
plt.ylabel('Loss')
plt.xlabel('Epoch')
plt.legend(['Train', 'Validation'], loc='upper right')

plt.tight_layout()
plt.savefig('training_history.png', dpi=300, bbox_inches='tight')
plt.close()

# 8. Make predictions for some test images
num_images_to_predict = 10
random_indices = np.random.choice(len(X_test), num_images_to_predict, replace=False)

# Get the images and their true labels
images_to_predict = X_test[random_indices]
true_labels = y_test[random_indices]

# Make predictions
predictions = model.predict(images_to_predict)
predicted_labels = np.argmax(predictions, axis=1)

# Display the images and predictions
plt.figure(figsize=(15, 8))
for i in range(num_images_to_predict):
    plt.subplot(2, 5, i + 1)
    plt.imshow(images_to_predict[i].reshape(28, 28), cmap=plt.cm.binary)
    color = 'green' if predicted_labels[i] == true_labels[i] else 'red'
    plt.title(f"Pred: {class_names[predicted_labels[i]]}\nTrue: {class_names[true_labels[i]]}", color=color)
    plt.axis('off')
plt.suptitle('Model Predictions', fontsize=16)
plt.tight_layout()
plt.subplots_adjust(top=0.88)
plt.savefig('model_predictions.png', dpi=300, bbox_inches='tight')
plt.close()

# 9. Generate confusion matrix
y_pred = model.predict(X_test)
y_pred_classes = np.argmax(y_pred, axis=1)

# Create and plot confusion matrix
cm = confusion_matrix(y_test, y_pred_classes)
plt.figure(figsize=(12, 10))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', xticklabels=class_names, yticklabels=class_names)
plt.xlabel('Predicted')
plt.ylabel('True')
plt.title('Confusion Matrix')
plt.savefig('confusion_matrix.png', dpi=300, bbox_inches='tight')
plt.close()

# 10. Generate classification report
print("\nClassification Report:")
report = classification_report(y_test, y_pred_classes, target_names=class_names)
print(report)

# Save the model
model.save('fashion_mnist_cnn_model.h5')
print("Model saved as 'fashion_mnist_cnn_model.h5'")


# 11. Class for detailed prediction on specific images
class FashionPredictor:
    def __init__(self, model_path='fashion_mnist_cnn_model.h5'):
        self.model = keras.models.load_model(model_path)
        self.class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
                            'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

    def predict_specific_image(self, image_idx, test_images, test_labels):
        """Make a prediction for a specific image from the test set."""
        if image_idx >= len(test_images):
            raise ValueError(f"Image index {image_idx} is out of bounds.")

        # Get the image and reshape it
        image = test_images[image_idx].reshape(1, 28, 28, 1)
        true_label = test_labels[image_idx]

        # Make prediction
        prediction = self.model.predict(image, verbose=0)
        predicted_class = np.argmax(prediction[0])
        confidence = prediction[0][predicted_class] * 100

        # Display image and prediction
        plt.figure(figsize=(6, 6))
        plt.imshow(image.reshape(28, 28), cmap=plt.cm.binary)

        # Set title color based on correctness
        color = 'green' if predicted_class == true_label else 'red'

        plt.title(f"Prediction: {self.class_names[predicted_class]} ({confidence:.2f}%)\n"
                  f"True: {self.class_names[true_label]}", color=color)

        plt.axis('off')
        plt.tight_layout()

        # Save and display the result
        plt.savefig(f'specific_prediction_{image_idx}.png', dpi=300, bbox_inches='tight')
        plt.close()

        # Return the results
        return {
            'image_idx': image_idx,
            'true_class': self.class_names[true_label],
            'predicted_class': self.class_names[predicted_class],
            'confidence': confidence,
            'correct': predicted_class == true_label
        }


# Create predictor instance
predictor = FashionPredictor('fashion_mnist_cnn_model.h5')

# Predict two specific images as required by the assignment
print("\nMaking predictions for two specific images:")
image_indices = [42, 123]  # Choose two images for detailed prediction
results = []

for idx in image_indices:
    result = predictor.predict_specific_image(idx, X_test, y_test)
    results.append(result)
    print(f"Image {idx}:")
    print(f"  True class: {result['true_class']}")
    print(f"  Predicted class: {result['predicted_class']}")
    print(f"  Confidence: {result['confidence']:.2f}%")
    print(f"  Correct: {result['correct']}")
    print()


# 12. Additional visualization: Activation Maps
def visualize_activations(image_idx):
    """Visualize the activation maps for a specific image."""
    if image_idx >= len(X_test):
        raise ValueError(f"Image index {image_idx} is out of bounds.")

    # Get the image
    image = X_test[image_idx:image_idx + 1]

    # Create a model that returns the activations of the first convolutional layer
    activation_model = models.Model(
        inputs=model.input,
        outputs=[
            model.layers[0].output,  # First conv layer
            model.layers[2].output  # Second conv layer
        ]
    )

    # Get activations
    activations = activation_model.predict(image)

    # Show input image
    plt.figure(figsize=(15, 8))
    plt.subplot(1, 3, 1)
    plt.imshow(image[0].reshape(28, 28), cmap='gray')
    plt.title(f"Input Image\n({class_names[y_test[image_idx]]})")
    plt.axis('off')

    # Show activations from first conv layer
    plt.subplot(1, 3, 2)
    plt.imshow(np.mean(activations[0][0], axis=-1), cmap='viridis')
    plt.title("Activation Map - Conv Layer 1")
    plt.axis('off')

    # Show activations from second conv layer
    plt.subplot(1, 3, 3)
    plt.imshow(np.mean(activations[1][0], axis=-1), cmap='viridis')
    plt.title("Activation Map - Conv Layer 2")
    plt.axis('off')

    plt.tight_layout()
    plt.savefig(f'activation_maps_{image_idx}.png', dpi=300, bbox_inches='tight')
    plt.close()


# Visualize activations for the two specific images
#for idx in image_indices:
#    visualize_activations(idx)

print("All visualizations have been saved as PNG files.")
print("Assignment completed successfully!")